lib/carrierwave/uploader/versions.rb

```diff
@@ -213,23 +213,60 @@ def url(*args)
       # Recreate versions and reprocess them. This can be used to recreate
       # versions if their parameters somehow have changed.
       #
-      def recreate_versions!(*versions)
+      def recreate_versions!(*names)
         # Some files could possibly not be stored on the local disk. This
         # doesn't play nicely with processing. Make sure that we're only
         # processing a cached file
         #
         # The call to store! will trigger the necessary callbacks to both
         # process this version and all sub-versions
-        if versions.any?
-          file = sanitized_file if !cached?
-          store_versions!(file, versions)
+
+        if names.any?
+          file = sanitized_file
+          set_versions_to_cache_and_store(names)
+          cache!(file)
+          store!(file)
```

Comment:
Now we avoid duplication and possible errors by using similar approach to cache/store in both branches of a condition.

Reply:
We noticed this change causes an AWS S3-backed store to trigger callbacks and send DELETES.  Is this expected behavior?

Reply:
Seems related to https://github.com/carrierwaveuploader/carrierwave/issues/2385

Reply:
Found the reason -- this change caused the cache storage to match the uploader:

https://github.com/carrierwaveuploader/carrierwave/commit/629afecbaeccd2300e4660b78ee36bd95dd845c5